File: app/src/main/java/com/colombia/credit/view/email/EmailFilter.java
package com.colombia.credit.view.email;

import android.widget.Filter;

import java.util.ArrayList;

/**
 * Created by weisl on 2019/6/6.
 */
public class EmailFilter extends Filter {

    private ArrayList<String> mEmailList;
    private IFilterResultListener mListener;

    public EmailFilter(ArrayList<String> emailList) {
        this.mEmailList = emailList;
    }

    @Override
    protected FilterResults performFiltering(CharSequence constraint) {
        FilterResults results = new FilterResults();
        if (constraint == null || constraint.length() == 0) {
            results.values = new ArrayList<EmailInputBean>();
            results.count = 0;
        } else {
            String inputText = constraint.toString();
            ArrayList<EmailInputBean> newList = new ArrayList<>();
            int lastIndex = inputText.lastIndexOf("@");
            EmailInputBean emailInputBean;
            StringBuilder sb = new StringBuilder();
            if (lastIndex == -1) {
                for (String s : mEmailList) {
                    emailInputBean = new EmailInputBean(inputText,
                            sb.append(getSubText(inputText)).append(s).toString());
                    newList.add(emailInputBean);
                    sb.delete(0, sb.length());
                }
            } else {
                String text = inputText.substring(lastIndex);
                for (String s : mEmailList) {
                    if (s.startsWith(text)) {
                        emailInputBean = new EmailInputBean(inputText,
                                sb.append(getSubText(inputText)).append(s).toString());
                        newList.add(emailInputBean);
                        sb.delete(0, sb.length());
                    }
                }
            }
            results.values = newList;
            results.count = newList.size();
        }
        return results;
    }

    @Override
    protected void publishResults(CharSequence constraint, FilterResults results) {
        if (mListener != null) {
            ArrayList<EmailInputBean> list = new ArrayList<>();
            try {
                if (results != null && results.count > 0) {//有符合过滤规则的数据
                    ArrayList<EmailInputBean> tempList = (ArrayList<EmailInputBean>) results.values;
                    list.addAll(tempList);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            mListener.filterResult(constraint, list);
        }
    }


    private String getSubText(String text) {
        int indexOf = text.indexOf("@");
        if (indexOf == -1) {
            return text;
        }
        return text.substring(0, indexOf);
    }

    public interface IFilterResultListener {
        void filterResult(CharSequence constraint, ArrayList<EmailInputBean> results);
    }

    public void setIFilterResultListener(IFilterResultListener listener) {
        this.mListener = listener;
    }

}

File: app/src/main/java/com/colombia/credit/view/email/EmailFilter.kt
package com.colombia.credit.view.email

import android.widget.Filter

class EmailFilter constructor(private val emailList: ArrayList<String>?): Filter() {

    private var mListener: IFilterResultListener? = null

    override fun performFiltering(constraint: CharSequence?): FilterResults {
        val results = FilterResults()
        if (constraint == null || constraint.isEmpty()) {
            results.values = ArrayList<EmailInputBean>()
            results.count = 0
        } else {
            val inputText = constraint.toString()
            val newList = ArrayList<EmailInputBean>()
            val lastIndex = inputText.lastIndexOf("@")
            var emailInputBean: EmailInputBean
            val sb = StringBuilder()
            if (lastIndex == -1) {
                val list = emailList
                if (list != null) {
                    for (s in list) {
                        emailInputBean = EmailInputBean(
                            inputText,
                            sb.append(getSubText(inputText)).append(s).toString()
                        )
                        newList.add(emailInputBean)
                        sb.delete(0, sb.length)
                    }
                }
            } else {
                val text = inputText.substring(lastIndex)
                val list = emailList
                if (list != null) {
                    for (s in list) {
                        if (s.startsWith(text)) {
                            emailInputBean = EmailInputBean(
                                inputText,
                                sb.append(getSubText(inputText)).append(s).toString()
                            )
                            newList.add(emailInputBean)
                            sb.delete(0, sb.length)
                        }
                    }
                }
            }
            results.values = newList
            results.count = newList.size
        }
        return results
    }

    override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
        if (mListener != null) {
            val list = ArrayList<EmailInputBean>()
            try {
                if (results != null && results.count > 0) { //有符合过滤规则的数据
                    val tempList = results.values as? ArrayList<EmailInputBean> ?:return
                    list.addAll(tempList)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            mListener?.filterResult(constraint, list)
        }
    }


    private fun getSubText(text: String): String? {
        val indexOf = text.indexOf("@")
        return if (indexOf == -1) {
            text
        } else text.substring(0, indexOf)
    }

    interface IFilterResultListener {
        fun filterResult(constraint: CharSequence?, results: ArrayList<EmailInputBean>?)
    }

    fun setIFilterResultListener(listener: IFilterResultListener?) {
        mListener = listener
    }
}